File: python/aoc/aoc2015.py
import unittest
import typing
import functools
import enum


class Token(enum.Enum):
    Open = '('
    Close = ')'


def one(input: str) -> tuple[int, typing.Optional[int]]:
    def parse_char(c: str) -> Token:
        if c == '(':
            return Token.Open
        elif c == ')':
            return Token.Close
        else:
            raise ValueError(f"picoprobpy: invalid char {c}")

    def solve(acc: tuple[int, typing.Optional[int]], next: tuple[int, Token]) -> tuple[int, typing.Optional[int]]:
        floor, first_floor_neg = acc
        i, token = next
        match token:
            case Token.Open:
                return (floor + 1, first_floor_neg)
            case Token.Close:
                if first_floor_neg == None and floor == 0:
                    return (floor - 1, i+1)
                else:
                    return (floor - 1, first_floor_neg)

    output = functools.reduce(
        solve,  # 2. solve: (int * int) * Token -> int
        enumerate(map(parse_char, list(input))),  # 1. parse_char: str -> Token
        (0, None),
    )

    return output


class TestAOC2015(unittest.TestCase):
    def test(self):
        self.assertEqual(one("(())")[0], 0)
        self.assertEqual(one("()()")[0], 0)
        self.assertEqual(one("(((")[0], 3)
        self.assertEqual(one("(()(()(")[0], 3)
        self.assertEqual(one("))(((((")[0], 3)
        self.assertEqual(one("())")[0], -1)
        self.assertEqual(one("))(")[0], -1)
        self.assertEqual(one(")))")[0], -3)
        self.assertEqual(one(")())())")[0], -3)
        self.assertEqual(one(")")[1], 1)
        self.assertEqual(one("()())")[1], 5)


if __name__ == '__main__':
    input = open("../../rust/src/aoc/data/2015_1")
    output = one(input.read())
    print(output)
    unittest.main()

File: src/aoc/aoc_2015.rs
use std::convert::TryFrom;
use std::io;

enum Token {
    Open,
    Close,
}

impl TryFrom<char> for Token {
    type Error = ();

    fn try_from(c: char) -> Result<Self, Self::Error> {
        match c {
            '(' => Ok(Token::Open),
            ')' => Ok(Token::Close),
            _ => Err(()),
        }
    }
}

/// one takes in a list of up/down commands encoded as parenthesis
/// and produces (final floor * first floor to reach -1)
///
/// # Examples
/// ```
/// use picoprob::aoc::aoc_2015;
/// let input = ")))";
/// let output = aoc_2015::one(&input).unwrap();
/// assert_eq!(output, (-3, Some(1)));
/// ```
pub fn one(input: &str) -> Result<(i32, Option<usize>), io::Error> {
    // parse: &str -> Vec<Token>
    // count: Vec<Token> -> i32
    let output = input
        .chars()
        .map(|c| Token::try_from(c))
        .collect::<Result<Vec<_>, _>>()
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "picoprob: invalid data"))?
        .iter()
        .enumerate()
        .fold((0, None), |(floor, first_floor_neg), (i, cmd)| match cmd {
            Token::Open => (floor + 1, first_floor_neg),
            Token::Close => match first_floor_neg {
                Some(_) => (floor - 1, first_floor_neg),
                None => {
                    if floor == 0 {
                        (floor - 1, Some(i + 1)) // weird indexing
                    } else {
                        (floor - 1, None)
                    }
                }
            },
        });

    Ok(output)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_one() {
        let inputs = vec!["(())", "()()"];
        inputs.iter().for_each(|&input| {
            let out = one(input).unwrap();
            assert_eq!(out.0, 0);
        });
    }

    #[test]
    fn test_two() {
        let inputs = vec!["(((", "(()(()(", "))((((("];
        inputs.iter().for_each(|&input| {
            let out = one(input).unwrap();
            assert_eq!(out.0, 3);
        });
    }

    #[test]
    fn test_three() {
        let inputs = vec!["())", "))("];
        inputs.iter().for_each(|&input| {
            let out = one(input).unwrap();
            assert_eq!(out.0, -1);
        });
    }

    #[test]
    fn test_four() {
        let inputs = vec![")))", ")())())"];
        inputs.iter().for_each(|&input| {
            let out = one(input).unwrap();
            assert_eq!(out.0, -3);
        });
    }

    #[test]
    fn test_five() {
        let input = ")";
        let output = one(input).unwrap();
        assert_eq!(output.1, Some(1))
    }

    #[test]
    fn test_six() {
        let input = "()())";
        let output = one(input).unwrap();
        assert_eq!(output.1, Some(5))
    }
}
